runtime/compiler/z/codegen/S390PrivateLinkage.cpp

```diff
@@ -1649,6 +1659,140 @@ J9::Z::PrivateLinkage::buildNoPatchingVirtualDispatchWithResolve(TR::Node *callN
+   // Removing the pre dependencies from the postDeps
+   postDeps->setAddCursorForPre(0);
+   postDeps->setNumPreConditions(0, trMemory());
+   TR::LabelSymbol *cFlowRegionStart = generateLabelSymbol(cg());
+   cursor = generateS390LabelInstruction(cg(), TR::InstOpCode::LABEL, callNode, cFlowRegionStart, preDeps, cursor);
+
+   TR::Register      *slotClassReg = cg()->allocateRegister();
+   TR::Register      *slotMethodAddressReg = cg()->allocateRegister();
+   TR::RegisterPair  *slotClassMethodRegPair = cg()->allocateConsecutiveRegisterPair(slotMethodAddressReg, slotClassReg);
+   postDeps->addPostCondition(slotClassReg, TR::RealRegister::LegalEvenOfPair);
+   postDeps->addPostCondition(slotMethodAddressReg, TR::RealRegister::LegalOddOfPair);
+   postDeps->addPostCondition(slotClassMethodRegPair, TR::RealRegister::EvenOddPair);
+   postDeps->addPostConditionIfNotAlreadyInserted(vftReg, TR::RealRegister::AssignAny);
+   cursor = generateRXInstruction(cg(), TR::InstOpCode::LPQ, callNode, slotClassMethodRegPair,
+                                    generateS390MemoryReference(regEP, offsetof(ccInterfaceData, slot1Class), cg()), cursor);
+   cursor = generateRRInstruction(cg(), TR::InstOpCode::CGR, callNode, slotClassReg, vftReg, cursor);
+   cursor =generateS390BranchInstruction(cg(), TR::InstOpCode::BCR, callNode, TR::InstOpCode::COND_BE, slotMethodAddressReg, cursor);
+
+   cursor = generateRXInstruction(cg(), TR::InstOpCode::LPQ, callNode, slotClassMethodRegPair,
+                                    generateS390MemoryReference(regEP, offsetof(ccInterfaceData, slot2Class), cg()), cursor);
+   cursor = generateRRInstruction(cg(), TR::InstOpCode::CGR, callNode, slotClassReg, vftReg, cursor);
+   cursor =generateS390BranchInstruction(cg(), TR::InstOpCode::BCR, callNode, TR::InstOpCode::COND_BE, slotMethodAddressReg, cursor);
+   
+   cursor = generateRXInstruction(cg(), TR::InstOpCode::LPQ, callNode, slotClassMethodRegPair,
+                                    generateS390MemoryReference(regEP, offsetof(ccInterfaceData, slot3Class), cg()), cursor);
+   cursor = generateRRInstruction(cg(), TR::InstOpCode::CGR, callNode, slotClassReg, vftReg, cursor);
+   cursor =generateS390BranchInstruction(cg(), TR::InstOpCode::BCR, callNode, TR::InstOpCode::COND_BE, slotMethodAddressReg, cursor);
+   
+   TR::LabelSymbol *interfaceCallHelperSnippetLabel = generateLabelSymbol(cg());
+   TR::S390InterfaceCallReadOnlySnippet *snippet = new (trHeapMemory()) TR::S390InterfaceCallReadOnlySnippet(
+      cg(),
+      callNode,
+      interfaceCallHelperSnippetLabel,
+      doneLabel,
+      interfaceDataAddress);
+   // Cache miss, going to slow path to lookup the interface method.
+   cursor = new (trHeapMemory()) TR::S390RILInstruction(TR::InstOpCode::BRCL, callNode, 0xF, snippet, NULL, cg());
+   cursor = new (trHeapMemory()) TR::S390NOPInstruction(TR::InstOpCode::NOP, 2, callNode, cg()); 
```

Comment:
Why is the NOP needed?

Reply:
According to comment in https://github.com/eclipse/openj9/blob/67d8cf0cd31d5215756678a1cd434fc04af98055/runtime/compiler/z/codegen/S390PrivateLinkage.cpp#L2189 It is needed for the patchupVirtualCall, I feel it needs more clear documentation then that. I remember, while testing when I commented out the NOP it was failing the tests, let me find-out the exact call path that that requires this NOP.

Reply:
Ok, we should put the same comment here as exists in the JVM path linked above. I could definitely see someone not knowing anything about this coming down and removing the NOP thinking it shouldn't exist.

Reply:
I have copied the same comment in https://github.com/eclipse/openj9/compare/e2eb175cf4e392b9a013648129c1209bc1790bc8..7e34301ef20a2e2b9b78a0d987f8c0e5eb801d0c 
Currently checking the path that needs this NOP, will modify if I find something useful.